api/src/main/java/com/velocitypowered/api/event/connection/ConnectionEstablishEvent.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2018-2021 Velocity Contributors
+ *
+ * The Velocity API is licensed under the terms of the MIT License. For more details,
+ * reference the LICENSE file in the api top-level directory.
+ */
+
+package com.velocitypowered.api.event.connection;
+
+import com.google.common.base.Preconditions;
+import com.velocitypowered.api.event.ResultedEvent;
+import com.velocitypowered.api.event.annotation.AwaitingEvent;
+import com.velocitypowered.api.proxy.InboundConnection;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Event called when a connection is initially established with the proxy.
+ * Velocity will wait for this event to fire before continuing with the connection.
+ */
+@AwaitingEvent
+public class ConnectionEstablishEvent implements ResultedEvent<ResultedEvent.GenericResult> {
+  private final InboundConnection connection;
+  private GenericResult result = GenericResult.allowed();
+
+  public ConnectionEstablishEvent(final @NotNull InboundConnection connection) {
+    this.connection = Preconditions.checkNotNull(connection, "connection");
+  }
+
+  /**
+   * Returns the inbound connection that is being established.
+   * @return the connection
+   */
+  public @NotNull InboundConnection getConnection() {
+    return this.connection;
+  }
+
+  @Override
+  public GenericResult getResult() {
+    return this.result;
+  }
+
+  @Override
+  public void setResult(final @NotNull GenericResult result) {
```

Comment:
```suggestion
  public void setResult(@NotNull GenericResult result) {
```
In other event classes, the method arguments are not declared final. In order to ensure a uniform code style, I would remove the final here.